README.md
[![npm version](https://badge.fury.io/js/%40mapbox%2Fsexagesimal.svg)](https://badge.fury.io/js/%40mapbox%2Fsexagesimal)
[![Build Status](https://travis-ci.org/mapbox/sexagesimal.svg)](https://travis-ci.org/mapbox/sexagesimal)

## sexagesimal

Convert between [sexagesimal coordinates](http://en.wikipedia.org/wiki/Sexagesimal) and
decimal coordinates.

### Usage

with npm (and/or) browserify

    npm install @mapbox/sexagesimal

otherwise

    curl https://raw.github.com/mapbox/sexagesimal/master/sexagesimal.js

### Examples

```js
var sexagesimal = require('sexagesimal');

// Converting a single DMS coordinate:

sexagesimal('40° 42′ 45.72″ N');  // direction after
// 40.712700000000005
sexagesimal('N40° 42′ 45.72″');   // direction before
// 40.712700000000005
sexagesimal('N40°42′45.72″');     // flexible whitespace
// 40.712700000000005

// Converting a coordinate pair from DMS to lat/lng:

sexagesimal.pair('40° 42′ 45.72″ N, 74° 0′ 21.24″ W');   // direction after
// [ 40.712700000000005, -74.0059 ]
sexagesimal.pair('N 40° 42′ 45.72″, W 74° 0′ 21.24″');   // direction before
// [ 40.712700000000005, -74.0059 ]

// Converting a lat/lon coordinate to DMS:

sexagesimal.coordToDMS(40.71270000000000, 'lat');
// { whole: 40, minutes: 42, seconds: 45, dir: 'N' }
sexagesimal.coordToDMS(-74.0059, 'lon');
// { whole: 74, minutes: 0, seconds: 21, dir: 'W' }


```

### API

```js
sexagesimal(str, dims) // returns a number or null
```

`dims` is by default `NSEW` but can be other ordinal directions expressed
as a string of characters.

```js
sexagesimal.pair(str, dims) // returns [lat, lon] or null
```

`dims` is by default `NSEW` but can be other ordinal directions expressed
as a string of characters.

```js
sexagesimal.format(float, dimension) // returns a formatted string
```

Format a single sexagesimal number. `dimension` must be a string, either
`"lat"` or `"lon"`.

```js
sexagesimal.formatPair({ lat: float, lon: float }) // returns a formatted string
```

Format a sexagesimal coordinate.


CHANGELOG.md
:warning: = breaking change

## 1.2.1
##### 2019-Sep-27
* Add support for format using "deg", such as found in some exif data [@counterbeing]

## 1.2.0
##### 2019-Aug-05
* Handle strings with preceding direction and no comma ([#8])
* Handle more variations of degree symbols ([#9], thanks [@quincylvania])
* Require Node 8+

[#8]: https://github.com/mapbox/sexagesimal/pull/8
[#9]: https://github.com/mapbox/sexagesimal/pull/9
[@quincylvania]: https://github.com/quincylvania

## 1.1.0
##### 2017-Aug-22
* Handle strings where direction appears before the coordinates ([#7], thanks [@sabas])

[#7]: https://github.com/mapbox/sexagesimal/pull/7
[@sabas]: https://github.com/sabas

## 1.0.0
##### 2017-Feb-13
* Bump all dependencies
* Require Node 4+
* :warning: sexagesimal is now a scoped package under @mapbox namespace

## 0.5.0
##### 2015-Aug-07
* Exposes `coordToDMS` method ([#5], thanks [@sabas])

[#5]: https://github.com/mapbox/sexagesimal/pull/5
[@sabas]: https://github.com/sabas

## 0.4.0
##### 2014-Dec-06
* Add `sexagesimal` binary. ([#4], thanks [@kapadia])

[#4]: https://github.com/mapbox/sexagesimal/pull/4
[@kapadia]: https://github.com/kapadia

## 0.3.0
##### 2014-Mar-17
* Added `.format(float, str) -> str` and `.formatPair({ lat: float, lon: float }) -> str` methods
  that generate sexagesimal strings from decimal values.

## 0.2.1
##### 2014-Jan-29
* No longer tolerates junk at the end of coordinate strings, which has led
  to false-positives like `10 12th Street`. ([#2], thanks [@yhahn])

[#2]: https://github.com/mapbox/sexagesimal/pull/2
[@yhahn]: https://github.com/yhahn


test/pair.js
var sexagesimal = require('../');
var test = require('tap').test;

test('plain ordinals', function(t) {
    t.deepEqual(sexagesimal.pair('66 32'), [66, 32]);
    t.deepEqual(sexagesimal.pair('66, 32'), [66, 32]);
    t.deepEqual(sexagesimal.pair('-66 -32'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('-66, -32'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('66° 30′ 360″, -66° 30′ 720"'), [66.6, -66.7]);
    t.end();
});

test('trailing dimension, no separator', function(t) {
    t.deepEqual(sexagesimal.pair('66N 32W'), [66, -32]);
    t.deepEqual(sexagesimal.pair('66 S 32 W'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('66S 32E'), [-66, 32]);
    t.deepEqual(sexagesimal.pair('32 w 66 s'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('66° 30′ 360″ N 66° 30′ 720" W'), [66.6, -66.7]);
    t.deepEqual(sexagesimal.pair('32W 66S'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('32 e 66 n'), [66, 32]);
    t.end();
});

test('trailing dimension, comma separator', function(t) {
    t.deepEqual(sexagesimal.pair('66N, 32W'), [66, -32]);
    t.deepEqual(sexagesimal.pair('66 S, 32 W'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('66S, 32E'), [-66, 32]);
    t.deepEqual(sexagesimal.pair('32 w , 66 s'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('66° 30′ 360″ N, 66° 30′ 720" W'), [66.6, -66.7]);
    t.deepEqual(sexagesimal.pair('32W , 66S'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('32 e, 66 n'), [66, 32]);
    t.end();
});

test('leading dimension, no separator', function(t) {
    t.deepEqual(sexagesimal.pair('N66 W32'), [66, -32]);
    t.deepEqual(sexagesimal.pair('S 66 W 32'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('S66 E32'), [-66, 32]);
    t.deepEqual(sexagesimal.pair('w 32 s 66'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('N 66° 30′ 360″  W 66° 30′ 720"'), [66.6, -66.7]);
    t.deepEqual(sexagesimal.pair('W32 S66'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('e 32 n 66'), [66, 32]);
    t.end();
});

test('leading dimension, comma separator', function(t) {
    t.deepEqual(sexagesimal.pair('N66, W32'), [66, -32]);
    t.deepEqual(sexagesimal.pair('S 66, W 32'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('S66, E32'), [-66, 32]);
    t.deepEqual(sexagesimal.pair('w 32 , s 66'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('N 66° 30′ 360″,  W 66° 30′ 720"'), [66.6, -66.7]);
    t.deepEqual(sexagesimal.pair('W32 , S66'), [-66, -32]);
    t.deepEqual(sexagesimal.pair('e 32, n 66'), [66, 32]);
    t.end();
});

test('returns null for non sexagesimal strings', function(t) {
    t.deepEqual(sexagesimal.pair('W32, S66 cruft'), null);
    t.deepEqual(sexagesimal.pair('cruft W32, S66'), null);
    t.deepEqual(sexagesimal.pair('cruft W32, S66 cruft'), null);
    t.deepEqual(sexagesimal.pair('500 johnson st 11310'), null);
    t.deepEqual(sexagesimal.pair('500 5th'), null);
    t.deepEqual(sexagesimal.pair('500 5th st nw'), null);
    t.deepEqual(sexagesimal.pair(' '), null);
    t.end();
});


test/parse.js
var sexagesimal = require('../'),
    test = require('tap').test;

test('basic directions with degrees', function(t) {
    t.deepEqual(sexagesimal('0° N'), 0);
    t.deepEqual(sexagesimal('66° N'), 66);
    t.deepEqual(sexagesimal('66º N'), 66);
    t.deepEqual(sexagesimal('66˚ N'), 66);
    t.deepEqual(sexagesimal('66.5° N'), 66.5);
    t.deepEqual(sexagesimal('66.2525° N'), 66.2525);
    t.deepEqual(sexagesimal('66° S'), -66);
    t.deepEqual(sexagesimal('16° W'), -16);
    t.deepEqual(sexagesimal('26° E'), 26);
    t.deepEqual(sexagesimal('N 66.5°'), 66.5);
    t.deepEqual(sexagesimal('W16°'), -16);
    t.deepEqual(sexagesimal('66 deg N'), 66);
    t.end();
});

test('decimals', function(t) {
    t.deepEqual(sexagesimal('26'), 26);
    t.deepEqual(sexagesimal('26.5'), 26.5);
    t.deepEqual(sexagesimal('26.52'), 26.52);
    t.deepEqual(sexagesimal('-26.52'), -26.52);
    t.end();
});

test('minutes', function(t) {
    t.deepEqual(sexagesimal('66° 30\' N'), 66.5);
    t.deepEqual(sexagesimal('66° 30\' E'), 66.5);
    t.deepEqual(sexagesimal('66° 30\' S'), -66.5);
    t.deepEqual(sexagesimal('66° 30\' W'), -66.5);
    t.deepEqual(sexagesimal('0° 30\' N'), 0.5);
    t.deepEqual(sexagesimal('66 deg 30\' N'), 66.5);
    t.end();
});

test('seconds', function(t) {
    t.deepEqual(sexagesimal('66° 30′ 360″ N'), 66.6);
    t.deepEqual(sexagesimal('66° 30′ 720″ S'), -66.7);
    t.deepEqual(sexagesimal('66° 30′ 720" S'), -66.7);
    t.deepEqual(sexagesimal('66° 30′ 720" E'), 66.7);
    t.deepEqual(sexagesimal('66° 30′ 720" W'), -66.7);
    t.deepEqual(sexagesimal('66 deg 30\' 720" N'), 66.7);
    t.end();
});

test('spaces', function(t) {
    t.deepEqual(sexagesimal('0°\tN'), 0);
    t.deepEqual(sexagesimal('66°  N'), 66);
    t.deepEqual(sexagesimal('0°N'), 0);
    t.end();
});


test/format.js
var sexagesimal = require('../'),
    test = require('tap').test;

test('format', function(t) {
    t.equal(sexagesimal.format(10, 'lat'), '10° N');
    t.equal(sexagesimal.format(10, 'lon'), '10° E');
    t.equal(sexagesimal.format(-10, 'lon'), '10° W');
    t.equal(sexagesimal.format(-10.5, 'lon'), '10° 30\' W');
    t.equal(sexagesimal.format(10.5, 'lon'), '10° 30\' E');
    t.equal(sexagesimal.format(10.51, 'lon'), '10° 30\' 35" E');
    t.equal(sexagesimal.format(10.51, 'lat'), '10° 30\' 35" N');
    t.end();
});

test('format & parse', function(t) {
    t.equal(sexagesimal(sexagesimal.format(10.51, 'lat')).toFixed(2), "10.51");
    t.equal(sexagesimal(sexagesimal.format(10.51, 'lon')).toFixed(2), "10.51");
    t.equal(sexagesimal(sexagesimal.format(10, 'lon')).toFixed(2), "10.00");
    t.equal(sexagesimal(sexagesimal.format(10.5, 'lon')).toFixed(2), "10.50");
    t.end();
});


index.js
module.exports = element;
module.exports.pair = pair;
module.exports.format = format;
module.exports.formatPair = formatPair;
module.exports.coordToDMS = coordToDMS;


function element(input, dims) {
  var result = search(input, dims);
  return (result === null) ? null : result.val;
}


function formatPair(input) {
  return format(input.lat, 'lat') + ' ' + format(input.lon, 'lon');
}


// Is 0 North or South?
function format(input, dim) {
  var dms = coordToDMS(input, dim);
  return dms.whole + '° ' +
    (dms.minutes ? dms.minutes + '\' ' : '') +
    (dms.seconds ? dms.seconds + '" ' : '') + dms.dir;
}


function coordToDMS(input, dim) {
  var dirs = { lat: ['N', 'S'], lon: ['E', 'W'] }[dim] || '';
  var dir = dirs[input >= 0 ? 0 : 1];
  var abs = Math.abs(input);
  var whole = Math.floor(abs);
  var fraction = abs - whole;
  var fractionMinutes = fraction * 60;
  var minutes = Math.floor(fractionMinutes);
  var seconds = Math.floor((fractionMinutes - minutes) * 60);

  return {
    whole: whole,
    minutes: minutes,
    seconds: seconds,
    dir: dir
  };
}


function search(input, dims) {
  if (!dims) dims = 'NSEW';
  if (typeof input !== 'string') return null;

  input = input.toUpperCase();
  var regex = /^[\s\,]*([NSEW])?\s*([\-|\—|\―]?[0-9.]+)(?:[°º˚]|\sDEG)?\s*(?:([0-9.]+)['’′‘]\s*)?(?:([0-9.]+)(?:''|"|”|″)\s*)?([NSEW])?/;

  var m = input.match(regex);
  if (!m) return null;  // no match

  var matched = m[0];

  // extract dimension.. m[1] = leading, m[5] = trailing
  var dim;
  if (m[1] && m[5]) {                 // if matched both..
    dim = m[1];                       // keep leading
    matched = matched.slice(0, -1);   // remove trailing dimension from match
  } else {
    dim = m[1] || m[5];
  }

  // if unrecognized dimension
  if (dim && dims.indexOf(dim) === -1) return null;

  // extract DMS
  var deg = m[2] ? parseFloat(m[2]) : 0;
  var min = m[3] ? parseFloat(m[3]) / 60 : 0;
  var sec = m[4] ? parseFloat(m[4]) / 3600 : 0;
  var sign = (deg < 0) ? -1 : 1;
  if (dim === 'S' || dim === 'W') sign *= -1;

  return {
    val: (Math.abs(deg) + min + sec) * sign,
    dim: dim,
    matched: matched,
    remain: input.slice(matched.length)
  };
}


function pair(input, dims) {
  input = input.trim();
  var one = search(input, dims);
  if (!one) return null;

  input = one.remain.trim();
  var two = search(input, dims);
  if (!two || two.remain) return null;

  if (one.dim) {
    return swapdim(one.val, two.val, one.dim);
  } else {
    return [one.val, two.val];
  }
}


function swapdim(a, b, dim) {
  if (dim === 'N' || dim === 'S') return [a, b];
  if (dim === 'W' || dim === 'E') return [b, a];
}


sexagesimal.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.sexagesimal = f()}})(function(){var define,module,exports;return (function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
module.exports = element;
module.exports.pair = pair;
module.exports.format = format;
module.exports.formatPair = formatPair;
module.exports.coordToDMS = coordToDMS;


function element(input, dims) {
  var result = search(input, dims);
  return (result === null) ? null : result.val;
}


function formatPair(input) {
  return format(input.lat, 'lat') + ' ' + format(input.lon, 'lon');
}


// Is 0 North or South?
function format(input, dim) {
  var dms = coordToDMS(input, dim);
  return dms.whole + '° ' +
    (dms.minutes ? dms.minutes + '\' ' : '') +
    (dms.seconds ? dms.seconds + '" ' : '') + dms.dir;
}


function coordToDMS(input, dim) {
  var dirs = { lat: ['N', 'S'], lon: ['E', 'W'] }[dim] || '';
  var dir = dirs[input >= 0 ? 0 : 1];
  var abs = Math.abs(input);
  var whole = Math.floor(abs);
  var fraction = abs - whole;
  var fractionMinutes = fraction * 60;
  var minutes = Math.floor(fractionMinutes);
  var seconds = Math.floor((fractionMinutes - minutes) * 60);

  return {
    whole: whole,
    minutes: minutes,
    seconds: seconds,
    dir: dir
  };
}


function search(input, dims) {
  if (!dims) dims = 'NSEW';
  if (typeof input !== 'string') return null;

  input = input.toUpperCase();
  var regex = /^[\s\,]*([NSEW])?\s*([\-|\—|\―]?[0-9.]+)(?:[°º˚]|\sDEG)?\s*(?:([0-9.]+)['’′‘]\s*)?(?:([0-9.]+)(?:''|"|”|″)\s*)?([NSEW])?/;

  var m = input.match(regex);
  if (!m) return null;  // no match

  var matched = m[0];

  // extract dimension.. m[1] = leading, m[5] = trailing
  var dim;
  if (m[1] && m[5]) {                 // if matched both..
    dim = m[1];                       // keep leading
    matched = matched.slice(0, -1);   // remove trailing dimension from match
  } else {
    dim = m[1] || m[5];
  }

  // if unrecognized dimension
  if (dim && dims.indexOf(dim) === -1) return null;

  // extract DMS
  var deg = m[2] ? parseFloat(m[2]) : 0;
  var min = m[3] ? parseFloat(m[3]) / 60 : 0;
  var sec = m[4] ? parseFloat(m[4]) / 3600 : 0;
  var sign = (deg < 0) ? -1 : 1;
  if (dim === 'S' || dim === 'W') sign *= -1;

  return {
    val: (Math.abs(deg) + min + sec) * sign,
    dim: dim,
    matched: matched,
    remain: input.slice(matched.length)
  };
}


function pair(input, dims) {
  input = input.trim();
  var one = search(input, dims);
  if (!one) return null;

  input = one.remain.trim();
  var two = search(input, dims);
  if (!two || two.remain) return null;

  if (one.dim) {
    return swapdim(one.val, two.val, one.dim);
  } else {
    return [one.val, two.val];
  }
}


function swapdim(a, b, dim) {
  if (dim === 'N' || dim === 'S') return [a, b];
  if (dim === 'W' || dim === 'E') return [b, a];
}

},{}]},{},[1])(1)
});
